extensions/register.md

```diff
@@ -0,0 +1,147 @@
+`FAIL REGISTER COMPLETE_CONNECTION_REQUIRED`, it SHOULD make a second attempt after it receives the
+welcome message.
+
+    VERIFY <account> <code>
+    
+The `VERIFY` command completes a registration that required email verification.
+
+### Responses
+
+    REGISTER SUCCESS <account> <message>
+    VERIFY SUCCESS <account> <message>
+    
+Sent by the server when the `REGISTER` or `VERIFY` fully succeeded, and no
+further action is needed.
+Afterward, the client is assumed to be authenticated to the `<account>`
+as if it used SASL.
+
+    REGISTER VERIFICATION_REQUIRED <account> <message>
+    
+Sent by the server as a response to `REGISTER` when further action is required
+to complete registration of the `<account>`, as explain in the human-readable `<message>`.
+
+    FAIL REGISTER USERNAME_EXISTS <account> <message>
+
+Sent by the server as a response to `REGISTER` when the client tried to register
+while using a nick that is not available as a new account's name
+(for example, because someone already registered it).
+
+The server MUST NOT send this response before the client sent a `NICK` command.
+
+The client MAY try registering again later.
+Clients should not automatically retry immediately without changing their nickname
+or waiting.
+
+    FAIL REGISTER ALREADY_REGISTERED <account> <message>
+    FAIL VERIFY ALREADY_REGISTERED <account> <message>
+
+Sent by the server is the client is already authenticated.
+
+    FAIL REGISTER WEAK_PASSWORD <account> <message>
+
+Sent by the server if the password is considered too weak.
+
+The client MAY try registering again.
+
+    FAIL REGISTER UNACCEPTABLE_PASSWORD <account> <message>
+
+Sent by the server if the password is invalid for any reason other than weakness.
+
+The client MAY try registering again.
+
+    FAIL REGISTER INVALID_EMAIL <account> <message>
+
+Sent by the server if it cannot send emails to the provided address.
+
+    FAIL REGISTER UNACCEPTABLE_EMAIL <account> <message>
+
+Sent by the server if the email address is valid, but not available for account registration.
+
+    FAIL REGISTER COMPLETE_CONNECTION_REQUIRED
+    FAIL VERIFY COMPLETE_CONNECTION_REQUIRED
+
+Sent by the server if the client sent a `REGISTER` command before connection registration.
+The server MUST NOT sent these replies if it advertizes the `before-connect` key of the
+`draft/register` capability.
+
+    FAIL VERIFY INVALID_CODE <account> <message>
+
+Sent by the server if the `<code>` in the `VERIFY` command is invalid or expired.
+
+    FAIL REGISTER TEMPORARILY_UNAVAILABLE <account> <message>
+    FAIL VERIFY TEMPORARILY_UNAVAILABLE <account> <message>
+
+Sent by the server if the `REGISTER`/`VERIFY` commands are temporarily unavailable.
+
+
+# Client considerations
+
+This section is non-normative.
+
+Passwords are opaque byte strings.
+It is recommended for them to be valid UTF-8;
+or authentication may be impossible later (eg. with SASL PLAIN).
```

Comment:
Pure pedantry: I don't think SASL PLAIN would have any problems here, since everything gets base64-encoded. The issue is more with being able to pass them easily to NickServ (e.g., to change the password).

I think a warning about password length may be in order? I've been imposing a cap of 300 bytes to ensure that passwords fit comfortably into IRC lines (it's theoretically possible that someone could submit a passphrase through this framework that gets truncated in other contexts). And although I haven't been validating this until now, it might also be worth checking that passphrases are acceptable as non-final IRC parameters (nonempty, no spaces, no initial `:`).

Reply:
Purer pedantry: RFC 4616 requires the encoded bytes to be a UTF-8 string.

IMO warnings about password length should be left to the UI. I could mention it in a non-normative section though.